### PR TITLE
Bandchain.js: Implement Msg

### LINKED ISSUE #2818: Implement MsgRequest on Client

### PR DESCRIPTION
**Pull request by Pitchanai**
Fixed: #2818 

## Implementation details
- Implement MsgRequest
- Implement MsgDelegate

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [x] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [x] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: helpers/bandchain2.js/src/message.ts [hunk 1]
```diff
@@ -1,11 +1,77 @@
...
+    calldata: string,
+    askCount: number,
+    minCount: number,
+    clientID: string,
+    sender: Address,
+  ) {
+    super()
+
+    this.oracleScriptID = oracleScriptID
+    this.calldata = Buffer.from(calldata, 'hex')
```
Comment (evilpeach) [inline helpers/bandchain2.js/src/message.ts]: User should pass `calldata` as Buffer type instead of hex string

**Review by evilpeach — CHANGES_REQUESTED** [review]

**Review by pzshine — APPROVED** [review]

File: helpers/bandchain2.js/src/message.ts [hunk 2]
```diff
@@ -43,3 +109,39 @@ export class MsgSend extends Msg {
...
+    }
+  }
+
+  getSender() {
+    return this.delegatorAddress
+  }
+
+  validate() {
+    // TODO: Uncomment this when coin.validate() is ready
+    // this.amount.validate()
```
Comment (evilpeach) [inline helpers/bandchain2.js/src/message.ts]: Let's validate it both MsgSend and MsgDelegate.

**Review by evilpeach — CHANGES_REQUESTED** [review]

File: helpers/bandchain2.js/src/data.ts [hunk 3]
```diff
@@ -9,9 +9,16 @@ export class Coin {
     this.denom = denom
   }
 
-  asJson() {
+  asJson(): object {
```
Comment (evilpeach) [inline helpers/bandchain2.js/src/data.ts]:
```
asJson(): {amount: string, denom: string} {
...
}
```

**Review by evilpeach — CHANGES_REQUESTED** [review]

**Review by pzshine — APPROVED** [review]

File: helpers/bandchain2.js/__tests__/message.test.ts [hunk 4]
```diff
@@ -1,12 +1,75 @@
...
-const { MsgSend } = Message
+const { MsgSend, MsgRequest, MsgDelegate } = Message
 const { Coin } = Data
 
-describe('MsgSend', () => {
-  const coin = new Coin(100000, 'uband')
+const coin = new Coin(100000, 'uband')
+
+describe('MsgRequest', () => {
+  const sender_addr = Address.fromAccBech32(
```
Comment (evilpeach) [inline helpers/bandchain2.js/__tests__/message.test.ts]: camel case

File: helpers/bandchain2.js/__tests__/message.test.ts [hunk 5]
```diff
@@ -1,12 +1,75 @@
...
+    'band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c',
+  )
+
+  const calldata = Buffer.from('000000034254430000000000000001', 'hex')
+  const memo = 'from_bandchain.js'
+
+  it('create successfully', () => {
+    const msgRequest = new MsgRequest(1, calldata, 2, 2, memo, sender_addr)
```
Comment (evilpeach) [inline helpers/bandchain2.js/__tests__/message.test.ts]: camel case

File: helpers/bandchain2.js/__tests__/message.test.ts [hunk 6]
```diff
@@ -1,12 +1,75 @@
...
+
+    msgs.forEach((msg, index) => {
+      expect(() => {
+        msg.validate()
+      }).toThrowError(errorText[index])
+    })
+  })
+})
 
+describe('MsgSend', () => {
```
Comment (evilpeach) [inline helpers/bandchain2.js/__tests__/message.test.ts:72]: fail case?

File: helpers/bandchain2.js/src/data.ts [hunk 7]
```diff
@@ -9,9 +9,16 @@ export class Coin {
...
+  asJson(): { amount: string; denom: string } {
     return { amount: this.amount.toString(), denom: this.denom }
   }
+
+  validate(): boolean {
+    if (this.amount < 0) throw Error('Expect amount more than 0')
+    if (this.denom.length == 0) throw Error('Expect denom')
+
+    return true
+  }
```
Comment (evilpeach) [inline helpers/bandchain2.js/src/data.ts]: create unit test for Coin

File: helpers/bandchain2.js/__tests__/message.test.ts [hunk 8]
```diff
@@ -21,5 +84,39 @@ describe('MsgSend', () => {
...
       },
     })
+    expect(msgSend.getSender().toAccBech32()).toEqual(
+      'band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c',
+    )
+    expect(msgSend.validate()).toBeTruthy()
+  })
+})
+
+describe('MsgDelegate', () => {
```
Comment (evilpeach) [inline helpers/bandchain2.js/__tests__/message.test.ts:106]: fail case?

File: helpers/bandchain2.js/src/message.ts [hunk 9]
```diff
@@ -1,11 +1,77 @@
...
+  getSender() {
+    return this.sender
+  }
+
+  validate() {
+    if (this.oracleScriptID <= 0)
+      throw Error('oracleScriptID cannot less than zero')
+    if (!Number.isInteger(this.oracleScriptID))
+      throw Error('oracleScriptID is not an integer')
+    if (this.calldata.length > constants.MAX_LENGTH)
```
Comment (evilpeach) [inline helpers/bandchain2.js/src/message.ts]: Let's use MAX_DATA_SIZE = 256

File: helpers/bandchain2.js/src/message.ts [hunk 10]
```diff
@@ -1,11 +1,77 @@
...
+      throw Error('too large calldata')
+    if (!Number.isInteger(this.askCount))
+      throw Error('askCount is not an integer')
+    if (!Number.isInteger(this.minCount))
+      throw Error('minCount is not an integer')
+    if (this.minCount <= 0)
+      throw Error(`invalid minCount got: minCount: ${this.minCount}`)
+    if (this.askCount < this.minCount)
+      throw Error(
+        `invalida askCount got: minCount: ${this.minCount}, askCount: ${this.askCount}`,
```
Comment (evilpeach) [inline helpers/bandchain2.js/src/message.ts]: typo -> invalida

**Review by evilpeach — CHANGES_REQUESTED** [review]

File: helpers/bandchain2.js/src/data.ts [hunk 11]
```diff
@@ -9,9 +9,17 @@ export class Coin {
...
 
-  asJson() {
+  asJson(): { amount: string; denom: string } {
     return { amount: this.amount.toString(), denom: this.denom }
   }
+
+  validate(): boolean {
+    if (!Number.isInteger(this.amount)) throw Error('amount is not an integer')
+    if (this.amount < 0) throw Error('Expect amount more than 0')
+    if (this.denom.length == 0) throw Error('Expect denom')
```
Comment (evilpeach) [inline helpers/bandchain2.js/src/data.ts]: ===

**Review by evilpeach — CHANGES_REQUESTED** [review]

**Review by evilpeach — APPROVED** [review]

**Pull request merged**